### ISSUE TITLE
💅 `lint/suspicious/noRedeclare` wrongly reported on double type inferrence

### ISSUE THREAD

**Issue by cyrilchapon** [biomejs/biome#7809]
### Environment information

- "@biomejs/biome": "2.2.6"

```bash
export type BaseColumnKey<T extends readonly string[]> = T[number] extends
  | `-${infer Base}`
  | infer Base
  ? Base
  : never
```

### Rule name

`lint/suspicious/noRedeclare`

### Playground link

https://biomejs.dev/playground/?code=ZQB4AHAAbwByAHQAIAB0AHkAcABlACAAQgBhAHMAZQBDAG8AbAB1AG0AbgBLAGUAeQA8AFQAIABlAHgAdABlAG4AZABzACAAcgBlAGEAZABvAG4AbAB5ACAAcwB0AHIAaQBuAGcAWwBdAD4AIAA9ACAAVABbAG4AdQBtAGIAZQByAF0AIABlAHgAdABlAG4AZABzAAoAIAAgAHwAIABgAC0AJAB7AGkAbgBmAGUAcgAgAEIAYQBzAGUAfQBgAAoAIAAgAHwAIABpAG4AZgBlAHIAIABCAGEAcwBlAAoAIAAgAD8AIABCAGEAcwBlAAoAIAAgADoAIABuAGUAdgBlAHIA

### Expected result

This should not report this as a redeclare. (or should it ?)

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by taga3s** [comment]
Hi! I'd like to work on this issue.
Could you assign it to me?
